[Issue 11496][C++] Allow partitioned producers to start lazily

Vanlightly opened the pull request:
Fixes #11496 also matches part of PIP 79.

C++ implementation that closely matches the proposed Java client changes from reducing partitioned producer connections and lookups: PR 10279

### Motivation

Producers that send messages to partitioned topics start a producer per partition, even when using single partition routing. For topics that have the combination of a large number of producers and a large number of partitions, this can put strain on the brokers. With say 1000 partitions and single partition routing with non-keyed messages, 999 topic owner lookups and producer registrations are performed that could be avoided.

PIP 79 also describes this. I wrote this before realising that PIP 79 also covers this. This implementation can be reviewed and contrasted to the Java client implementation in https://github.com/apache/pulsar/pull/10279.

### Modifications

Allows partitioned producers to start producers for individual partitions lazily. Starting a producer involves a topic owner
lookup to find out which broker is the owner of the partition, then registering the producer for that partition with the owner
broker. For topics with many partitions and when using SinglePartition routing without keyed messages, all of these
lookups and producer registrations are a waste except for the single chosen partition.

This change allows the user to control whether a producer on a partitioned topic uses this lazy start or not, via a new config
in ProducerConfiguration. When ProducerConfiguration.setLazyStartPartitionedProducers(true) is set, the PartitionedProducerImpl.start() becomes a synchronous operation that only does housekeeping (no network operations).
The producer of any given partition is started (which includes a topic owner lookup and registration) upon sending the first message to that partition. While the producer starts, messages are buffered.

The sendTimeout timer is only activated once a producer has been fully started, which should give enough time for any buffered messages to be sent. For very short send timeouts, this setting could cause send timeouts during the start phase. The default of 30s should however not cause this issue.

### Verifying this change

This change added tests and can be verified as follows:
  - BasicEndToEndTest, testPartitionedProducerConsumer
  - BasicEndToEndTest, testSyncFlushBatchMessagesPartitionedTopicLazyProducers
  - BasicEndToEndTest, testFlushInLazyPartitionedProducer

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (yes) - client configuration
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

#### For contributor

For this PR, do we need to update docs?

Yes, the new client config would need documenting. Can contribute that if this PR is accepted.


#### For committer

For this PR, do we need to update docs?

- If yes,
  
  - if you update docs in this PR, label this PR with the `doc` label.
  
  - if you plan to update docs later, label this PR with the `doc-required` label.

  - if you need help on updating docs, create a follow-up issue with the `doc-required` label.
  
- If no, label this PR with the `no-need-doc` label and explain why.

ivankelly commented on pulsar-client-cpp/lib/PartitionedProducerImpl.cc:
```diff
@@ -87,13 +87,18 @@ unsigned int PartitionedProducerImpl::getNumPartitionsWithLock() const {
-        std::bind(&PartitionedProducerImpl::handleSinglePartitionProducerCreated,
-                  const_cast<PartitionedProducerImpl*>(this)->shared_from_this(), _1, _2, partition));
+
+    if (conf_.getLazyStartPartitionedProducers()) {
```
This looks like a noop. Rather than having "createLazyPartitionProducer" count the number of producers created and change state when all done, you could change ::start() to change the state.

Vanlightly replied:
It's not a no-op as it kicks off the partition update task which is something we want running, even if no producers have started. It seems the most simple way of doing it.

ivankelly replied:
What I was suggesting was to kick off the update task in ::start() if lazyproducer is enabled. Is numProducersCreated used for anything else?

Vanlightly replied:
Yes, it's required when closing. Also, this method will be called when the partition update task runs and see a new number of partitions, so it fits well with that.

ivankelly replied:
:+1:

ivankelly commented on pulsar-client-cpp/lib/PartitionedProducerImpl.cc:
```diff
@@ -161,7 +182,13 @@ void PartitionedProducerImpl::sendAsync(const Message& msg, SendCallback callbac
     // find a producer for that partition, index should start from 0
     ProducerImplPtr producer = producers_[partition];
+
+    if (conf_.getLazyStartPartitionedProducers() && !producer->isStarted()) {
```
is checking the conf needed here?

Vanlightly replied:
No, it isn't, I'll remove that.

ivankelly commented on pulsar-client-cpp/lib/PartitionedProducerImpl.cc:
```diff
@@ -161,7 +182,13 @@ void PartitionedProducerImpl::sendAsync(const Message& msg, SendCallback callbac
     ProducerImplPtr producer = producers_[partition];
+
+    if (conf_.getLazyStartPartitionedProducers() && !producer->isStarted()) {
+        producer->start();
```
is producer::start() idempotent?

Vanlightly replied:
It isn't but this code is protected by the producersLock so it can't be called twice concurrently.

ivankelly commented on pulsar-client-cpp/lib/PartitionedProducerImpl.cc:
```diff
@@ -209,7 +236,10 @@ void PartitionedProducerImpl::closeAsync(CloseCallback closeCallback) {
     // when `state_` is Ready
     for (auto& producer : producers_) {
-        if (!producer->isClosed()) {
+        if (!producer->isStarted()) {
```
what happens if a client does a sendAsync concurrently with closeAsync?

Vanlightly replied:
I am working on some changes to ensure these can be called concurrently and sane things happen.

ivankelly commented on pulsar-client-cpp/lib/PartitionedProducerImpl.cc:
```diff
@@ -379,8 +418,8 @@ bool PartitionedProducerImpl::isConnected() const {
     producersLock.unlock();
-    for (const auto& producer : producers_) {
-        if (!producer->isConnected()) {
+    for (const auto& producer : producers) {
```
If the producer has been created, but no messages have been sent, this will return true, even though there has been no connections.

Vanlightly replied:
Yes. The question is what people do with this method. Is it an indicator of a problem? Without lazy producers, false means we were once connected but now aren't and so can be considered a transient problem. But with lazy producers, false may mean we've not even tried, so false does not mean anything. So I decided to return true to signal absence of a problem.

Definitely interested to hear opinions on that.

ivankelly replied:
Ya, I have no strong opinion here. "connected" is a weird concept in a distributed system like this.

ivankelly commented on pulsar-client-cpp/lib/ProducerImpl.cc:
```diff
@@ -550,10 +559,14 @@ void ProducerImpl::batchMessageTimeoutHandler(const boost::system::error_code& e
+
+    // ignore if the producer is already closing/closed
+    if (assertState(Ready)) {
+        Lock lock(mutex_);
```
should the assertState not happen inside the lock?

ivankelly replied:
I see that assertState take the lock, but there's a window between that returning and it being taken again where the producer could be closed.

Vanlightly replied:
I've fixed that and added some extra state_ checks before batchMessageAndSend to avoid race conditions.

Anonymitaet commented:
@Vanlightly Thanks for providing doc-related info! Looking forward to your docs.

BewareMyPower commented:
Could you rebase to master since C++/Python tests have been fixed?

Vanlightly commented:
I realised that I had not included the extra config in the __init__.py so will add that to this PR.

@Anonymitaet Beyond the doxygen and pdoc generated docs, is it worth documenting this change further. If so, where do you think?

Anonymitaet commented:
@Vanlightly how about adding your changes [here](https://pulsar.apache.org/docs/en/next/client-libraries-cpp/)?

Vanlightly commented:
@Anonymitaet I have added to the documentation a section about lazy producers and also added examples of both synchronous and asynchronous producers and consumers.

Anonymitaet commented on site2/docs/client-libraries-cpp.md:
```diff
@@ -253,52 +253,227 @@ pulsar+ssl://pulsar.us-west.example.com:6651
 
-To use Pulsar as a consumer, you need to create a consumer on the C++ client. The following is an example. 
+To use Pulsar as a consumer, you need to create a consumer on the C++ client. There are two main ways of using the consumer:
+- blocking style: synchronously calling `receive(msg)`
+- asynchronous (event based) style: using a message listener
```
capitalize the first word?

Anonymitaet commented on site2/docs/client-libraries-cpp.md:
```diff
@@ -253,52 +253,227 @@ pulsar+ssl://pulsar.us-west.example.com:6651
 
-To use Pulsar as a producer, you need to create a producer on the C++ client. The following is an example. 
+To use Pulsar as a producer, you need to create a producer on the C++ client. There are two main ways of using a producer:
+- blocking style where each call to `send` waits for an ack from the broker.
+- non-blocking asynchronous style where `sendAsync` is called instead of `send` and a callback is supplied for when the ack is received from the broker.
```
same comment as previously

Anonymitaet commented on site2/docs/client-libraries-cpp.md:
````diff
@@ -253,52 +253,227 @@ pulsar+ssl://pulsar.us-west.example.com:6651
+}
+```
+
+### Non-blockging example
````
blocking?

Anonymitaet commented on site2/docs/client-libraries-cpp.md:
```diff
@@ -253,52 +253,227 @@ pulsar+ssl://pulsar.us-west.example.com:6651
+
+The producer configuration `blockIfQueueFull` is useful here to avoid `ResultProducerQueueIsFull` errors when the internal queue for outgoing send requests becomes full. Once the internal queue is full, `sendAsync` becomes blocking which can make your code simpler.
+
+Without this configuration the `Result` in the callback is the error `ResultProducerQueueIsFull` and you must decide how to deal with that.
```
```suggestion
Without this configuration, the `Result` in the callback returns the error `ResultProducerQueueIsFull`. You must decide how to deal with that.
```
do you mean this?

Anonymitaet commented on site2/docs/client-libraries-cpp.md:
````diff
@@ -253,52 +253,227 @@ pulsar+ssl://pulsar.us-west.example.com:6651
-client.close();
+```
+
+### Partitioned Topics and Lazy Producers
````
```suggestion
### Partitioned topics and lazy producers
```
Pulsar docs use sentence case for headings

Anonymitaet commented on site2/docs/client-libraries-cpp.md:
```diff
@@ -253,52 +253,227 @@ pulsar+ssl://pulsar.us-west.example.com:6651
+
+### Partitioned Topics and Lazy Producers
+
+When scaling out a Pulsar topic, you may configure a topic to have hundreds of partitions. Likewise, you may have also scaled out your producers so there are hundreds or even thousands of producers. This can put some strain the the Pulsar brokers as when you create a producer on a partitioned topic, internally it creates one internal producer per partition which involves communications to the bro ...
```
```suggestion
When scaling out a Pulsar topic, you may configure a topic to have hundreds of partitions. Likewise, you may have also scaled out your producers so there are hundreds or even thousands of producers. This can put some strain the Pulsar brokers as when you create a producer on a partitioned topic, internally it creates one internal producer per partition which involves communications to the brokers for each one. So for a topic with 1000 partitions and 1000 producers, it ends up creating 1,000,000 internal producers across the producer applications, each of which has to communicate with a broker to find out which broker it should connect to and then perform the connection handshake.
```

Anonymitaet commented on site2/docs/client-libraries-cpp.md:
```diff
@@ -253,52 +253,227 @@ pulsar+ssl://pulsar.us-west.example.com:6651
+
+When scaling out a Pulsar topic, you may configure a topic to have hundreds of partitions. Likewise, you may have also scaled out your producers so there are hundreds or even thousands of producers. This can put some strain the the Pulsar brokers as when you create a producer on a partitioned topic, internally it creates one internal producer per partition which involves communications to the bro ...
+
+You can reduce the load caused by this combination of large number of partitions and many producers by doing the following:
+- use SinglePartition partition routing mode (this ensures that all messages are only sent to a single, randomly selected partition)
+- use non-keyed messages (when messages are keyed, routing is based on the hash of the key and so messages will end up being sent to multiple partitions)
+- use lazy producers (this ensures that an internal producer is only created on demand when a message needs to be routed to a partition)
```
```suggestion
You can reduce the load caused by this combination of a large number of partitions and many producers by doing the following:
- use SinglePartition partition routing mode (this ensures that all messages are only sent to a single, randomly selected partition)
- use non-keyed messages (when messages are keyed, routing is based on the hash of the key and so messages will end up being sent to multiple partitions)
- use lazy producers (this ensures that an internal producer is only created on demand when a message needs to be routed to a partition)
```

Anonymitaet commented on site2/docs/client-libraries-cpp.md:
```diff
@@ -253,52 +253,227 @@ pulsar+ssl://pulsar.us-west.example.com:6651
+
+With our example above, that reduces the number of internal producers spread out over the 1000 producer apps from 1,000,000 to just 1000.
+
+Note that it does mean that there can be extra latency for the first message sent. If you have a low send timeout set, this could trigger if the initial connection handshake is slow to complete.
```
```suggestion
Note that there can be extra latency for the first message sent. If you set a low send timeout, this could trigger if the initial connection handshake is slow to complete.
```
trigger what or sth is triggered？

Anonymitaet reviewed (commented):
Many thanks for your great contribution! I’ve left some comments, PTAL.

Vanlightly commented:
@Anonymitaet thanks, I agree with all comments/suggestions and I have fixed accordingly.

Anonymitaet reviewed (approved):
Thanks for your contribution. Approved from the doc perspective.

equanz commented:
Thank you for your comments about PIP-79.
I have a few comments. I think these comments don't block merging.

For authn/authz backward compatibility reason, I think the internal producer should be connected to broker before completing the creation of partitioned producer( https://mail-archives.apache.org/mod_mbox/pulsar-dev/202102.mbox/%3CCAO2yDybn3sqPJV32YqvYndk%3D8mxNKodcGB4GE3QmUs8F9m8YUw%40mail.gmail.com%3E ). However, it isn't critical because the change affects when ProducerConfiguration.setLazyStartPartitionedProducers(true) is set.

As mentioned in [here](https://github.com/apache/pulsar/wiki/PIP-79%3A-Reduce-redundant-producers-from-partitioned-producer#change-partitionedtopicstats-about-producer), partitioned producer stats will be incorrect if each partition has different number of producers. I tried to fix this issue in the PIP-79 (It will be fixed in https://github.com/apache/pulsar/pull/10534 at Java client).

ivankelly commented:
> For authn/authz backward compatibility reason, I think the internal producer should be connected to broker before completing the creation of partitioned producer( https://mail-archives.apache.org/mod_mbox/pulsar-dev/202102.mbox/%3CCAO2yDybn3sqPJV32YqvYndk%3D8mxNKodcGB4GE3QmUs8F9m8YUw%40mail.gmail.com%3E ). However, it isn't critical because the change affects when ProducerConfiguration.setLazyStartPartitionedProducers(true) is set.

This can be worked around by making the producer connection to any one random partition. If we could make this the first partition selected in roundrobin, even better.

Vanlightly commented:
@equanz @ivankelly I have changed it so that one producer is started eagerly, chosen by the routing policy and any other internal producers are started lazily. This will ensure that authz errors occur during createProducer and we still get the desired behaviour of fewer lookups and connections when using single partition routing.

BewareMyPower commented:
@codelipenghui I added the `release/2.8.2` label first.

Without this PR, the test introduced from https://github.com/apache/pulsar/pull/11955 would block forever. It's because for a partitioned producer, calling `send()` after  `shutdown()` or `close()` would block forever. I didn't find why this PR works, but this PR should fix it and has been tested in my local env.

The root cause is when a partitioned producer calls `shutdown()` in client shutdown, it only sets the state and the internal producers are not affected. We can also solve this problem by shutdown all internal producers when a partitioned producer calls `shutdown`.
